pkg/config/handlerConfigurer.go

```diff
@@ -0,0 +1,164 @@
+// Copyright 2017 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package config
+
+import (
+	"fmt"
+
+	"github.com/golang/glog"
+	"github.com/golang/protobuf/proto"
+
+	pbd "istio.io/api/mixer/v1/config/descriptor"
+	pb "istio.io/mixer/pkg/config/proto"
+	"istio.io/mixer/pkg/expr"
+	"istio.io/mixer/pkg/template"
+)
+
+type (
+	handlerConfigurer struct {
+		tmplRepo       template.Repository
+		typeChecker    expr.TypeChecker
+		attrDescFinder expr.AttributeDescriptorFinder
+	}
+	instancesByTemplate struct {
+		instancesNamesByTemplate map[string][]string
+	}
+)
+
+func (t *instancesByTemplate) insertInstance(tmplName string, instName string) {
+	instsPerTmpl, exists := t.instancesNamesByTemplate[tmplName]
+	if !exists {
+		t.instancesNamesByTemplate[tmplName] = make([]string, 0)
+	}
+
+	if !contains(instsPerTmpl, instName) {
+		t.instancesNamesByTemplate[tmplName] = append(t.instancesNamesByTemplate[tmplName], instName)
+	}
+}
+
+func newInstancesByTemplate() instancesByTemplate {
+	return instancesByTemplate{make(map[string][]string)}
+}
+
+func configureHandlers(actions []*pb.Action, constructors map[string]*pb.Constructor,
+	handlers map[string]*HandlerBuilderInfo, tmplRepo template.Repository, expr expr.TypeChecker, df expr.AttributeDescriptorFinder) error {
+	configurer := handlerConfigurer{tmplRepo: tmplRepo, typeChecker: expr, attrDescFinder: df}
+
+	iTypes, err := configurer.inferTypes(constructors)
+	if err != nil {
+		return err
+	}
+	grpHandlers, err := configurer.groupHandlerInstancesByTemplate(actions, constructors, handlers)
+	if err != nil {
+		return err
+	}
+
+	return configurer.dispatchTypesToHandlers(iTypes, grpHandlers, handlers)
+}
+
+func (h *handlerConfigurer) dispatchTypesToHandlers(infrdTypes map[string]proto.Message,
+	instsByTmpls map[string]instancesByTemplate, handlers map[string]*HandlerBuilderInfo) error {
+	for hName, instsByTmpl := range instsByTmpls {
+		hb, found := handlers[hName]
+		if !found {
+			// This should not happen, since the dispatchTypesToHandlers should get called after all verifications.
+			return fmt.Errorf("handler %s is not registered", hName)
+		}
+
+		for tmplName, insts := range instsByTmpl.instancesNamesByTemplate {
+			ti, found := h.tmplRepo.GetTemplateInfo(tmplName)
+			if !found {
+				// This should not happen, since the dispatchTypesToHandlers should get called after all verifications.
+				return fmt.Errorf("template %s is not registered", tmplName)
```

Comment:
well, these are errors from users which they can fix. However, based on the current code flow, this will not happen since we validate before we try to configure handler in this code. Unit test exercise this code.